kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/DekManager.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.dek;
+
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import io.kroxylicious.kms.service.Kms;
+import io.kroxylicious.kms.service.KmsService;
+import io.kroxylicious.kms.service.Serde;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * A DekManager encapsulates a {@link Kms}, providing access to the ability to
+ * encrypt and decrypt using key material from the KMS without exposing
+ * that key material outside this package.
+ * @param <K> The type of KEK id
+ * @param <E> The type of encrypted DEK
+ */
+@ThreadSafe
+public class DekManager<K, E> {
+
+    private final Kms<K, E> kms;
+    private final int maxEncryptionsPerDek;
+
+    public <C> DekManager(KmsService<C, K, E> kmsService, C config, int maxEncryptionsPerDek) {
```

Comment:
Maybe better to take KMS as parameter because we have external code to wrap the service provided KMS in decorating implementations that handle retry/caching.

Reply:
There's a couple of other ways we could handle that:

* Rather than having decorators for `Kms` we could just as well decorate the `DekManager`.
* Alternatively have a `KmsService` decorator which added the decorations to the Kms. 

To be fair, I'm not really sure whether we're really getting any additional security out of this arrangement. Anything with access to a `Kms` has access to all the KEKs that the KMS credentials have access to, so it feels like we should try to avoid things having direct access to a `Kms`. But the same is also true of something which can construct a KMS using a `KmsService` and a config for it. So I'm not really sure how much we're gaining here, though perhaps we could do something useful with `module` and `export ... to`.